src/machine/i2c_generic.go

```diff
@@ -0,0 +1,305 @@
+	data := (address << 1)
+	if !write {
+		data |= 1 // set read flag
+	}
+
+	i2c.scl.High()
+	i2c.sda.Low()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	for i := 0; i < 8; i++ {
+		i2c.scl.Low()
+		if ((data >> (7 - i)) & 1) == 1 {
+			i2c.sda.High()
+		} else {
+			i2c.sda.Low()
+		}
+		i2c.waitHalfCycle()
+		i2c.waitHalfCycle()
+		i2c.scl.High()
+		i2c.waitHalfCycle()
+		i2c.waitHalfCycle()
+	}
+
+	i2c.scl.Low()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.sda.Configure(PinConfig{Mode: PinInputPullup})
+	i2c.scl.High()
+	i2c.waitHalfCycle()
+
+	i2c.nack = i2c.sda.Get()
+
+	i2c.waitHalfCycle()
+
+	// wait until bus ready
+
+	return nil
+}
+
+func (i2c *SI2C) signalStop() error {
+	i2c.scl.Low()
+	i2c.sda.Low()
+	i2c.sda.Configure(PinConfig{Mode: PinOutput})
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.scl.High()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.sda.High()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	return nil
+}
+
+func (i2c *SI2C) signalRead() error {
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.scl.Low()
+	i2c.sda.Low()
+	i2c.sda.Configure(PinConfig{Mode: PinOutput})
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.scl.High()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	return nil
+}
+
+func (i2c *SI2C) readByte() byte {
+	var data byte
+	for i := 0; i < 8; i++ {
+		i2c.scl.Low()
+		i2c.sda.Configure(PinConfig{Mode: PinInputPullup})
+		i2c.waitHalfCycle()
+		i2c.waitHalfCycle()
+		i2c.scl.High()
+		if i2c.sda.Get() {
+			data |= 1 << (7 - i)
+		}
+		i2c.waitHalfCycle()
+		i2c.waitHalfCycle()
+	}
+	return data
+}
+
+func (i2c *SI2C) sendNack() error {
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.scl.Low()
+	i2c.sda.High()
+	i2c.sda.Configure(PinConfig{Mode: PinOutput})
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	i2c.scl.High()
+	i2c.waitHalfCycle()
+	i2c.waitHalfCycle()
+	return nil
+}
+
+// WriteRegister transmits first the register and then the data to the
+// peripheral device.
+//
+// Many I2C-compatible devices are organized in terms of registers. This method
+// is a shortcut to easily write to such registers. Also, it only works for
+// devices with 7-bit addresses, which is the vast majority.
+func (i2c *SI2C) WriteRegister(address uint8, register uint8, data []byte) error {
+	buf := make([]uint8, len(data)+1)
+	buf[0] = register
+	copy(buf[1:], data)
+	return i2c.Tx(uint16(address), buf, nil)
+}
+
+// ReadRegister transmits the register, restarts the connection as a read
+// operation, and reads the response.
+//
+// Many I2C-compatible devices are organized in terms of registers. This method
+// is a shortcut to easily read such registers. Also, it only works for devices
+// with 7-bit addresses, which is the vast majority.
+func (i2c *SI2C) ReadRegister(address uint8, register uint8, data []byte) error {
+	return i2c.Tx(uint16(address), []byte{register}, data)
+}
+
+func (i2c *SI2C) waitHalfCycle() {
+	for i := uint32(0); i < i2c.waitNum; i++ {
+		arm.Asm("nop")
```

Comment:
Now we have a dependency on device/arm.
To support RISCV, etc., you may want to change to a different method.

Reply:
The dependency on this part makes it arm-specific code.
Instead of i2c_generic.go, i2c_arm.go would be a better name.

Reply:
You can add architecture specific versions of `waitHalfCycle` in separate files and keep the rest of the code generic.